File: app/src/main/java/com/faciee/cti/valbastrelu/eticket/ui/bus/model/BusViewModel.java
package com.faciee.cti.valbastrelu.eticket.ui.bus.model;

import android.util.Log;

import androidx.lifecycle.LiveData;

import com.faciee.cti.valbastrelu.eticket.base.AbstractAndroidViewModel;
import com.faciee.cti.valbastrelu.eticket.base.ETicketApp;
import com.faciee.cti.valbastrelu.eticket.repo.ETkBusRepository;
import com.faciee.cti.valbastrelu.eticket.room.entities.Ticket;
import com.faciee.cti.valbastrelu.eticket.room.entities.Route;
import com.faciee.cti.valbastrelu.eticket.room.entities.Transaction;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public class BusViewModel extends AbstractAndroidViewModel {
	private static final String TAG = "BusActivityModel";
	
	private ETkBusRepository repository;
	
	private Ticket ticketActiv = null;
	LiveData<List<Ticket>> bilete;


	public BusViewModel(@NotNull ETicketApp application) {
		super(application);
		repository = application.getBusRepository();
		bilete = repository.getBilete();
	}
	
	public void setTicketActiv(Ticket ticketActiv) {
		this.ticketActiv = ticketActiv;
	}
	
	public LiveData<List<Ticket>> getLiveDataBilete(){
		return bilete;
	}
	
	public LiveData<List<Route>> getLiveDataTrasee(){
		return repository.getLiveDataTrasee();
	}
	
	public LiveData<List<String>> getLiveDataStatii(int nrTraseu){
		return repository.getLiveDataStatii(nrTraseu);
	}
	
	public LiveData<List<Transaction>> getLiveDataTranzactii(){
		return repository.getLiveDataTranzactii();
	}
	
	//INSERTS
	public void insertBilet(Ticket ticket){
		setTicketActiv(ticket);
		repository.insertBilet(ticket);
	}
	
	@Override
	protected void onCleared() {
		super.onCleared();
		Log.d(TAG, "onCleared: called");
	}
}

File: app/src/main/java/com/faciee/cti/valbastrelu/eticket/ui/bus/vm/BusViewModel.kt
package com.faciee.cti.valbastrelu.eticket.ui.bus.vm

import android.util.Log
import androidx.annotation.VisibleForTesting
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.faciee.cti.valbastrelu.eticket.base.AbstractAndroidViewModel
import com.faciee.cti.valbastrelu.eticket.base.ETicketApp
import com.faciee.cti.valbastrelu.eticket.repo.BusRepository
import com.faciee.cti.valbastrelu.eticket.room.entities.Route
import com.faciee.cti.valbastrelu.eticket.room.entities.Ticket
import com.faciee.cti.valbastrelu.eticket.room.entities.Transaction
import com.faciee.cti.valbastrelu.eticket.ui.common.TransportType
import kotlinx.coroutines.launch

class BusViewModel(
		application: ETicketApp
) : AbstractAndroidViewModel(application) {

	@VisibleForTesting
	constructor(app: ETicketApp, repository: BusRepository) : this(app) {
		this.repository = repository
	}

	private var repository: BusRepository
	private var activeTicket: Ticket? = null
	var busTicketsLiveData: LiveData<List<Ticket>>

	val liveDataTrasee: MutableLiveData<List<Route>> = MutableLiveData()


	init {
		repository = application.busRepository
		busTicketsLiveData = repository.ticketsLiveData

		viewModelScope.launch {
			liveDataTrasee.value = repository.getRoutes(TransportType.BUS)
		}
	}

	fun setTicketActiv(ticketActiv: Ticket?) {
		this.activeTicket = ticketActiv
	}

	fun getLiveDataStatii(nrTraseu: Int): LiveData<List<String>> {
		return repository.getLiveDataStatii(nrTraseu)
	}

	val liveDataTranzactii: LiveData<List<Transaction>>
		get() = repository.liveDataTranzactii

	fun saveTicket(ticket: Ticket) {
		setTicketActiv(ticket)
		viewModelScope.launch {
			repository.insertTicketInDatabase(ticket)
		}
	}

	override fun onCleared() {
		super.onCleared()
		Log.d(TAG, "onCleared: called")
	}

	companion object {
		private const val TAG = "BusActivityModel"

		fun getFactory(app: ETicketApp): ViewModelProvider.Factory {
			return object : ViewModelProvider.Factory {
				@Suppress("UNCHECKED_CAST")
				override fun <T : ViewModel?> create(modelClass: Class<T>): T {
					return BusViewModel(app) as T
				}
			}
		}
	}
}